mindmeld/components/question_answerer.py

```diff
@@ -449,805 +591,2172 @@ def _generate_mapping_data(embedder_model, embedding_fields):
+                    as the _ids, they will be attached to the curated results for corresponding _ids
+            Returns:
+                list[dict]: compiled docs
+            """
+            docs = {}
+
+            if _scores:
+                if len(_scores) != len(_ids):
+                    msg = f"Number of ids ({len(_ids)}) supplied did not match " \
+                          f"number of  scores ({len(_scores)}). Discarding inputted scores " \
+                          f"while curating docs for QA results. "
+                    logger.warning(msg)
+                else:
+                    docs = {_id: {"_id": _id, "_score": _scores[i]} for i, _id in enumerate(_ids)}
+
+            if not docs:
+                # when no _scores are inputted or when number of _scores do not match umber of _ids
+                docs = {_id: {"_id": _id} for _id in _ids}
+
+            # populate docs for all collected ids
+            for field_name, field_resource in index_resources.items():
+                for _id in _ids:
+                    try:
+                        docs[_id][field_name] = field_resource.id2value[_id]
+                    except KeyError:
+                        docs[_id][field_name] = None
+
+            return [*docs.values()]
+
+    ALL_INDICES = Indices()
+
+
+class QuestionAnswerer:
+
+    def __new__(cls, app_path=None, resource_loader=None, es_host=None, config=None, **kwargs):
+        """
+        This method is used to initialize a QuestionAnswerer based on model_type
+
+        To keep the code base backwards compatible, we use a '__new__()' way of creating instances
+        alongside using a factory approach. For cases wherein a question-answerer is instantiated
+        from 'QuestionAnswerer' class instead of  'QuestionAnswerer.create_question_answerer',
+        this method is called before __init__ and returns an instance of a question-answerer.
+
+        See that the input arguments are kept as-is with respect to the '__init__()' of
+        'ElasticsearchQuestionAnswerer' class which was the 'QuestionAnswerer' class in previous
+        version of 'question_answerer.py'
+        """
+
+        config = cls._get_config(config, app_path)
+        kwargs.update({
+            "app_path": app_path,
+            "resource_loader": resource_loader,
+            "es_host": es_host,
+            "config": config,
+        })
+        return cls._get_question_answerer(config)(**kwargs)
+
+    @staticmethod
+    def _get_config(config=None, app_path=None):
+        if not config:
+            return get_classifier_config("question_answering", app_path=app_path)
+        return config
+
+    @staticmethod
+    def _get_question_answerer(config):
+
+        # TODO: Both QA classes assume input text is in English.
+        #  Going forward, this should be configurable and multilingual support should be added!
+
+        use_elastic_search = config.get("use_elastic_search", True)
```

Comment:
i'm wondering if we should make this config more generic? This is related to another comment i have on the naming of the non-Elasticsearch based implementation. For example, if we were to add another implementation/approach this config param could become a bit confusing?

Reply:
update:
the configs will now be passed in as follows:
```
QUESTION_ANSWERER_CONFIG = {
   "model_type": <native/elasticsearch>
   "model_settings": {
       "query_type": "embedder_text",
       "embedder_type": "bert",
       "embedding_fields": {
           ...
       }
   }
}
```